sql/core/src/main/scala/org/apache/spark/sql/execution/commands/parsers.scala

```diff
@@ -0,0 +1,420 @@
+          val (cols, values, storedAsDirs) = skewedArgs match {
+            case Token("TOK_TABLESKEWED", skewedCols :: skewedValues :: stored) =>
+              val cols = skewedCols.children.map(n => unquoteString(cleanIdentifier(n.text)))
+              val values = skewedValues match {
+                case Token("TOK_TABCOLVALUE", values) =>
+                  Seq(values.map(n => unquoteString(cleanIdentifier(n.text))))
+                case Token("TOK_TABCOLVALUE_PAIR", pairs) =>
+                  pairs.map {
+                    case Token("TOK_TABCOLVALUES", values :: Nil) =>
+                      values match {
+                        case Token("TOK_TABCOLVALUE", vals) =>
+                          vals.map(n => unquoteString(cleanIdentifier(n.text)))
+                      }
+                  }
+              }
+
+              val storedAsDirs = stored match {
+                case Token("TOK_STOREDASDIRS", Nil) :: Nil => true
+                case _ => false
+              }
+
+              (cols, values, storedAsDirs)
+          }
+          (false, Some((cols, values, storedAsDirs)))
+      }
+
+      if (skewedArgs.isDefined) {
+        AlterTableSkewed(
+          tableIdent,
+          skewedArgs.get._1, /* cols */
+          skewedArgs.get._2, /* values */
+          skewedArgs.get._3, /* storedAsDirs */
+          notSkewed, notStoredAsDirs)(node.source)
+      } else {
+        AlterTableSkewed(tableIdent, Nil, Nil, false, notSkewed, notStoredAsDirs)(node.source)
+      }
+
+    case Token("TOK_ALTERTABLE_SKEWED_LOCATION", args) :: rest =>
+      val skewedMaps = args(0) match {
+        case Token("TOK_SKEWED_LOCATIONS", locationList :: Nil) =>
+          locationList match {
+            case Token("TOK_SKEWED_LOCATION_LIST", locationMaps) =>
+              locationMaps.map {
+                case Token("TOK_SKEWED_LOCATION_MAP", key :: value :: Nil) =>
+                  val k = key match {
+                    case Token(const, Nil) => Seq(unquoteString(cleanIdentifier(const)))
+                    case Token("TOK_TABCOLVALUES", values :: Nil) =>
+                      values match {
+                        case Token("TOK_TABCOLVALUE", vals) =>
+                          vals.map(n => unquoteString(cleanIdentifier(n.text)))
+                      }
+                  }
+                  (k, unquoteString(cleanIdentifier(value.text)))
+              }.toMap
+          }
+      }
+      AlterTableSkewedLocation(tableIdent, skewedMaps)(node.source)
+
+    case Token("TOK_ALTERTABLE_ADDPARTS", addPartsArgs) :: rest =>
+      val allowExisting = base.getClauseOption("TOK_IFNOTEXISTS", addPartsArgs)
+      val parts = if (allowExisting.isDefined) {
+        addPartsArgs.tail
+      } else {
+        addPartsArgs
+      }
+
+      val partitions: ArrayBuffer[(Map[String, Option[String]], Option[String])] =
+        new ArrayBuffer()
+      var currentPart: Map[String, Option[String]] = null
+      parts.map {
```

Comment:
You are using side effects, foreach is preferred. Is the problem you are trying to solve that partition locations are not nested?